test/recipes/70-test_dtlsrecords.t

```diff
@@ -0,0 +1,153 @@
+
+#Test 2: Injecting in context empty records should succeed
+$proxy->clear();
+$content_type = TLSProxy::Record::RT_HANDSHAKE;
+$inject_recs_num = 1;
+$proxy->serverflags("-min_protocol DTLSv1.2 -max_protocol DTLSv1.2");
+$proxy->clientflags("-max_protocol DTLSv1.2");
+$proxy->filter(\&add_empty_recs_filter);
+$proxy_start_success = $proxy->start();
+ok($proxy_start_success && TLSProxy::Message->success(), "In context empty records test");
+
+#Unrecognised record type tests
+
+#Test 3: Sending an unrecognised record type in DTLSv1.2 should fail
+$fatal_alert = 0;
+$proxy->clear();
+$proxy->serverflags("-min_protocol DTLSv1.2 -max_protocol DTLSv1.2");
+$proxy->clientflags("-max_protocol DTLSv1.2");
+$proxy->filter(\&add_unknown_record_type);
+ok($proxy->start() == 0, "Unrecognised record type in DTLS1.2");
+
+SKIP: {
+    skip "DTLSv1 disabled", 1 if disabled("dtls1");
+
+    #Test 4: Sending an unrecognised record type in DTLSv1 should fail
+    $fatal_alert = 0;
+    $proxy->clear();
+    $proxy->clientflags("-min_protocol DTLSv1 -max_protocol DTLSv1 -cipher DEFAULT:\@SECLEVEL=0");
+    $proxy->ciphers("AES128-SHA:\@SECLEVEL=0");
+    $proxy->filter(\&add_unknown_record_type);
+    ok($proxy->start() == 0, "Unrecognised record type in DTLSv1");
+}
+
+sub add_empty_recs_filter
+{
+    my $proxy = shift;
+    my $records = $proxy->record_list;
+
+    # We're only interested in the initial ClientHello
+    if ($proxy->flight != 0) {
+        $fatal_alert = 1 if @{$records}[-1]->is_fatal_alert(1) == TLSProxy::Message::AL_DESC_UNEXPECTED_MESSAGE;
+        return;
+    }
+
+    for (my $i = 0; $i < $inject_recs_num; $i++) {
+        my $record = TLSProxy::Record->new_dtls(
+            0,
+            $content_type,
+            TLSProxy::Record::VERS_TLS_1_2,
+            0,
+            0,
+            0,
+            0,
+            0,
+            0,
+            "",
+            ""
+        );
+        push @{$records}, $record;
+    }
+}
+
+sub add_unknown_record_type
+{
+    my $proxy = shift;
+    my $records = $proxy->record_list;
+    state $added_record;
+
+    # We'll change a record after the initial version neg has taken place
+    if ($proxy->flight == 0) {
+        $added_record = 0;
+        return;
+    } elsif ($proxy->flight != 1 || $added_record) {
+        $fatal_alert = 1 if @{$records}[-1]->is_fatal_alert(0) == TLSProxy::Message::AL_DESC_UNEXPECTED_MESSAGE;
+        return;
+    }
+
+    my $record = TLSProxy::Record->new_dtls(
+        1,
+        TLSProxy::Record::RT_UNKNOWN,
+        @{$records}[-1]->version(),
+        @{$records}[-1]->epoch(),
+        @{$records}[-1]->seq() +1,
+        1,
+        0,
+        1,
+        1,
+        "X",
+        "X"
+    );
+
+    #Find ServerHello record and insert after that
+    my $i;
+    for ($i = 0; ${$proxy->record_list}[$i]->flight() < 1; $i++) {
+        next;
+    }
+    $i++;
+
+    splice @{$proxy->record_list}, $i, 0, $record;
+    $added_record = 1;
+}
```

Comment:
This is basically a reduced version of the sslrecords test - with essentially identical filters except for new->new_dtls. Can we not find a way to reuse the existing script for DTLS in order to avoid all the copy&paste?

Reply:
Possibly. I'm just not that familiar with the perl testing framework yet. You are thinking something in the terms of

```
my @run_with_dtls = (0, 1);

foreach(@run_with_dtls) {
    run_test($_)
}
```

The next question from me would be: What is a nice way to skip the DTLS tests on windows (Since I couldn't get it to work on windows)?

Reply:
Something along those lines.

> The next question from me would be: What is a nice way to skip the DTLS tests on windows (Since I couldn't get it to work on windows)?

You can test for windows via `if $^O eq 'MSWin32'`

You can skip individual tests like this:

https://github.com/openssl/openssl/blob/0f644b96d209443b4566f7e86e3be2568292e75b/test/recipes/70-test_key_share.t#L142-L151

Reply:
Cool, I'll give it a go